ex-4-Linh/cap.c

```diff
@@ -0,0 +1,933 @@
+#include <time.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/time.h>
+#include <sys/socket.h>
+#include <netdb.h>
+#include <netinet/ip.h>
+#include <netinet/udp.h>
+#include <netinet/ip_icmp.h>
+#include <netinet/tcp.h>
+#include <netinet/in.h>
+#include <ifaddrs.h>
+#include <linux/if_ether.h>
+#include <arpa/inet.h>
+#include <unistd.h>
+#include <linux/if.h>
+#include <linux/if_packet.h>
+#include <sys/ioctl.h>
+
+#define ETH_PROTOCOL_IP4 0x0800
+
+#define IP_PROTOCOL_TCP 6
+#define IP_PROTOCOL_ICMP 1
+#define IP_PROTOCOL_IGMP 2
+#define IP_PROTOCOL_UDP 17
+
+#define MAX_FRAME_SIZE 100000
+
+#define ANY_CHAR '*'
+
+int IsInteger(char* s){
+    int i = -1;
+    
+    while (s[++i] != 0) {
+        if (s[i] < '0' || s[i] > '9')
+            return 0;
+    }
+    
+    return 1;
+}
+
+int IsIPAddress(char* ip){
+    unsigned int digit[4] = {0, 0, 0, 0};
+    int i = -1;
+    int length = 0;
+    int newDigit = 1;
+    
+    if (ip[0] == '\0') {
+        return 0;
+    }
+    
+    while (ip[++i]) {
+        if (ip[i] >= '0' && ip[i] <= '9') {
+            digit[length] *= 10;
+            digit[length] += ip[i] - '0';
+            if (digit[length] > 255) {
+                return 0;
+            }
+            newDigit = 0;
+        }
+        else if (ip[i] == '.') {
+            if (newDigit) {
+                return 0;
+            }
+            
+            length++;
+            newDigit = 1;
+            
+            if (length >= 4) {
+                return 0;
+            }
+        }
+        else {
+            return 0;
+        }
+    }
+    
+    if (length == 3){
+        return 1;
+    }
+    else {
+        return 0;
+    }
+}
+
+int CompareIPAddress(char* a, char* b){
```

Comment:
Acceptable. But you can use simple way by storing them as binary and use "&" to compare.

Reply:
Em nghĩ không phải &.
Mà là ^ (XOR).
if (a1 ^ a2 == 0) {//equals} else {//not equals}

Reply:
Nhưng function này mục đích của em là compare pattern IP